### PR TITLE
[JENKINS-41516] Add script console listener

### PR DESCRIPTION
**Pull request by daniel-beck**
Alternative to #6539 (but might still get reintegrated there) by @meiswjn, for now just to get a PR build/incremental deployment.

See [JENKINS-41516](https://issues.jenkins.io/browse/JENKINS-41516).

Minor secondary TODOs:

- [ ] Add implementation of extension point outside core (IMO kinda optional since `DefaultScriptListener` exists)

### Proposed changelog entries

- Developer: Add extension point to notify about in-process scripting events.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7056"></a>

### CODE REVIEW

**Comment by meiswjn** [comment]
Hi Daniel,
Thanks for your investment into this topic!
Do you want to implement this PR or https://github.com/jenkinsci/jenkins/pull/6539? For me, both is okay. I am just interested in having the feature available soon :D

**Comment by daniel-beck** [comment]
@meiswjn FYI I've finished this PR. I have no preference about proceeding with this PR, or integrating the changes here into your PR. Let me know if you want to do the latter.

**Comment by meiswjn** [comment]
Hey @daniel-beck, thank you! I have no preference either, I am just here for the features 😄 I think it is easiest to keep this one 👍

**Comment by meiswjn** [comment]
Tagging the reviewers that approved the previous PR #6539: @NotMyFault @res0nance @timja @StefanSpieker

**Comment by meiswjn** [comment]
@daniel-beck perhaps you can tag the reviewer group? :)

**Review by timja — APPROVED** [review]
nice

**Review by StefanSpieker — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
> Tagging the reviewers that approved the previous PR #6539: NotMyFault @res0nance timja StefanSpieker

My humble apology, I must have missed this PR :(

**Comment by meiswjn** [comment]
No worries, thanks for the approval! Looking forward to the release 🚀

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by NotMyFault** [comment]
The tests added in this PR cause failures in the default branch: https://ci.jenkins.io/job/Core/job/jenkins/job/master/5460/testReport/jenkins.model/

Would you mind taking a look and think you can submit a fixup before next Tuesday? Otherwise, I'd revert the merge to restore the test functionality and not break the weekly release.

**Comment by daniel-beck** [comment]
@NotMyFault Any idea why CI passed for this PR?

**Comment by daniel-beck** [comment]
https://github.com/jenkinsci/jenkins/blob/778b8b62de4e6f42da87cd78dbe8d731a6888cf6/Jenkinsfile#L109 

So we're running every test on every JDK on Linux 3 times, but less than a third of tests on Windows?

**Comment by daniel-beck** [comment]
How do I even confirm a fix PR will work, if this just skips execution of the test I'm fixing?

(Also, seems like a clear bug to not execute _new_ tests.)

**Comment by NotMyFault** [comment]
> https://github.com/jenkinsci/jenkins/blob/778b8b62de4e6f42da87cd78dbe8d731a6888cf6/Jenkinsfile#L109
> 
> So we're running every test on Linux 3 times (11, 17, and 19), but less than a third of tests on Windows?

But that's only used for the launchable part, isn't it https://github.com/jenkinsci/jenkins/blob/778b8b62de4e6f42da87cd78dbe8d731a6888cf6/Jenkinsfile#L113-L116 ?

> @NotMyFault Any idea why CI passed for this PR?

To be fair, it passed at the end of July, I could've updated this PR before integrating it into master.

**Comment by daniel-beck** [comment]
> But that's only used for the launchable part, isn't it 

Unsure what you mean. https://ci.jenkins.io/job/Core/job/jenkins/job/PR-7056/12/consoleFull did not run the new test on Windows, just three times on Linux.

**Comment by timja** [comment]
> How do I even confirm a fix PR will work, if this just skips execution of the test I'm fixing?
> 
> (Also, seems like a clear bug to not execute _new_ tests.)

cc @basil if you have any input on the above ^^

**Comment by daniel-beck** [comment]
> How do I even confirm a fix PR will work, if this just skips execution of the test I'm fixing?

Given that it's being executed in an unrelated PR now, this comment is probably obsolete.

**Comment by MarkEWaite** [comment]
Thanks for reporting the issue.

Launchable was enabled months ago to reduce the cost of pull request evaluation.  Launchable uses results from previous tests to define a subset of Windows tests that will execute in roughly the same amount of time as the Linux builds require.  The full set of Windows tests are run on the master branch after the pull request is merged.

This is the first time since Launchable was enabled that we've had a test failure "escape" from a pull request merge and reach the master branch.  I think that the cost savings from limiting Windows tests to approximately one hour (instead of the [four hours required on the main branch](https://ci.jenkins.io/job/Core/job/jenkins/job/master/5453/pipeline-console/?selected-node=217)) are worth this type of rare event.

I agree that the test selection algorithm used by Launchable should add all new tests to the list of tests that it runs.  We'll need to report that to them, since others will likely want the same thing.  When there is no test result data for a test, add it to the executed set of tests.

**Comment by MarkEWaite** [comment]
> I agree that the test selection algorithm used by Launchable should add all new tests to the list of tests that it runs. We'll need to report that to them, since others will likely want the same thing. When there is no test result data for a test, add it to the executed set of tests.

I've reported the issue to Launchable support and copied Kohsuke on the report so that he's aware.  Thanks again for detecting the issue!

File: core/src/main/java/jenkins/util/ScriptListener.java
Comment (jglick) [inline core/src/main/java/jenkins/util/ScriptListener.java:133]: This is implicitly `public`. If you did you not mean for this to be part of the API, use `@Restricted(NoExternalUse.class)`. Ditto other members of the interface.

File: core/src/main/java/jenkins/util/DefaultScriptListener.java
Comment (jglick) [inline core/src/main/java/jenkins/util/DefaultScriptListener.java:40]: Should be omitted because it is not part of the API anyway.

File: core/src/main/java/jenkins/util/ScriptListener.java
Comment (daniel-beck) [reply core/src/main/java/jenkins/util/ScriptListener.java:133]: Could go either way. Realistically there are few (possibly none) possible callers of `#fire…` methods outside core (that possibly could want this), so it might be better to start off `@Restricted` and open up if necessary.

File: core/src/main/java/jenkins/util/DefaultScriptListener.java
Comment (daniel-beck) [reply core/src/main/java/jenkins/util/DefaultScriptListener.java:40]: I like it to quickly see when a feature was added. While I could get there via `git blame`, given the negligible cost, this is reasonable to have IMO.

**Comment by basil** [comment]
Causes [JENKINS-72181](https://issues.jenkins.io/browse/JENKINS-72181).

**Pull request merged**
